src/manager/monitor_proxy.c

```diff
@@ -0,0 +1,262 @@
+        if (subscription == NULL) {
+                return NULL;
+        }
+
+        subscription->monitor = proxy_monitor_ref(monitor);
+        subscription->free_monitor = (free_func_t) proxy_monitor_unref;
+
+        subscription->handle_unit_new = proxy_monitor_on_unit_new;
+        subscription->handle_unit_removed = proxy_monitor_on_unit_removed;
+        subscription->handle_unit_state_changed = proxy_monitor_on_unit_state_changed;
+        subscription->handle_unit_property_changed = proxy_monitor_on_unit_property_changed;
+
+        return subscription;
+}
+
+ProxyMonitor *proxy_monitor_new(
+                Manager *manager, Node *requesting_node, const char *proxy_object_path, Node *providing_node) {
+        _cleanup_proxy_monitor_ ProxyMonitor *monitor = malloc0(sizeof(ProxyMonitor));
+        if (monitor == NULL) {
+                return NULL;
+        }
+
+        monitor->ref_count = 1;
+        monitor->manager = manager;
+        monitor->requesting_node = requesting_node;
+        monitor->providing_node = providing_node;
+
+        monitor->proxy_object_path = strdup(proxy_object_path);
+        if (monitor->proxy_object_path == NULL) {
+                return NULL;
+        }
+
+        return steal_pointer(&monitor);
+}
+
+ProxyMonitor *proxy_monitor_ref(ProxyMonitor *monitor) {
+        monitor->ref_count++;
+        return monitor;
+}
+
+void proxy_monitor_unref(ProxyMonitor *monitor) {
+        monitor->ref_count--;
+        if (monitor->ref_count != 0) {
+                return;
+        }
+
+        free(monitor->proxy_object_path);
+        free(monitor);
+}
+
+void proxy_monitor_close(ProxyMonitor *monitor) {
+        Manager *manager = monitor->manager;
+
+        if (monitor->subscription != NULL) {
+                manager_remove_subscription(manager, monitor->subscription);
+                subscription_unref(monitor->subscription);
+        }
+}
+
+int proxy_monitor_on_unit_new(
+                UNUSED void *userdata,
+                UNUSED const char *node,
+                UNUSED const char *unit,
+                UNUSED const char *reason) {
+        hirte_log_debug("ProxyMonitor processing unit new");
+        return 0;
+}
+
+int proxy_monitor_on_unit_state_changed(
+                UNUSED void *userdata,
+                UNUSED const char *node,
+                UNUSED const char *unit,
+                UNUSED const char *active_state,
+                UNUSED const char *substate,
+                UNUSED const char *reason) {
+        hirte_log_debug("ProxyMonitor processing unit state changed");
+        ProxyMonitor *monitor = (ProxyMonitor *) userdata;
+
+        int r = 0;
+        UnitActiveState state = active_state_from_string(active_state);
+        if (state == UNIT_ACTIVE || state == UNIT_FAILED) {
+                r = proxy_monitor_send_state(monitor, state);
+                proxy_monitor_close(monitor);
+                manager_remove_proxy_monitor(monitor->manager, monitor);
+        }
```

Comment:
It is a bit weird that we track failure to activate in two places, both in the proxy monitor, as well as in the Job callback (proxy_start_unit_callback). Its quite possible that you report the same thing twice, or that something else causes a state_change before the actual job completes. 

I think the proper way to handle this is to start the job on new proxy request, and only reply to that request with the resulting status from the job request. The monitor should only be uses to track the state of the unit after the job is finished. Then when that reports the unit as stopped we send a message to the agent to stop the proxy.

Reply:
Hmm, that would require not only calling StartUnit, but also tracking the job that results from that.
Also, would need the job monitoring to be robust against agent disconnect. So, maybe using the monitor to track start is a better approach anyway?

But if so, I wonder if the current approach is robust enough. It does nothing until the state changes to active or failed, but maybe there are other possible scenarios. Maybe it was it "deactivating" state when the monitor was created, and then it goes into inactive state. Although in that case the StartJob should probably trigger a new start. I dunno, it feels like there are potential problems here.